OpenRA.Game/Network/Order.cs

```diff
@@ -247,91 +265,116 @@ public Order(string orderString, Actor subject, Target target, bool queued)
 
 		public byte[] Serialize()
 		{
-			var minLength = OrderString.Length + 1 + (IsImmediate ? 1 + 1 + TargetString.Length + 1 + 4 : 6);
+			var minLength = 1 + OrderString.Length + 1;
+			if (Type == OrderType.Handshake)
+				minLength += TargetString.Length + 1;
+			else if (Type == OrderType.Immediate)
+				minLength += 1 + (TargetString != null ? TargetString.Length + 1 : 0) + (ExtraData != 0 ? 4 : 0);
+			else if (Type == OrderType.World)
+				minLength += 4 + 1 + 13 + (TargetString != null ? TargetString.Length + 1 : 0) + 4 + 4;
+
+			// ProtocolVersion.Orders and the associated documentation MUST be updated if the serialized format changes
 			var ret = new MemoryStream(minLength);
 			var w = new BinaryWriter(ret);
 
-			OrderFields fields = 0;
-			if (TargetString != null)
-				fields |= OrderFields.TargetString;
+			w.Write((byte)Type);
+			w.Write(OrderString);
+			switch (Type)
+			{
+				case OrderType.Handshake:
+				{
+					w.Write(TargetString);
 
-			if (ExtraData != 0)
-				fields |= OrderFields.ExtraData;
+					break;
+				}
 
-			if (IsImmediate)
-			{
-				w.Write((byte)0xFE);
-				w.Write(OrderString);
-				w.Write((byte)fields);
+				case OrderType.Immediate:
+				{
+					var fields = OrderFields.None;
+					if (TargetString != null)
+						fields |= OrderFields.TargetString;
 
-				if (fields.HasField(OrderFields.TargetString))
-					w.Write(TargetString);
+					if (ExtraData != 0)
+						fields |= OrderFields.ExtraData;
 
-				if (fields.HasField(OrderFields.ExtraData))
-					w.Write(ExtraData);
+					w.Write((byte)fields);
 
-				return ret.ToArray();
-			}
+					if (fields.HasField(OrderFields.TargetString))
+						w.Write(TargetString);
 
-			w.Write((byte)0xFF);
-			w.Write(OrderString);
-			w.Write(UIntFromActor(Subject));
+					if (fields.HasField(OrderFields.ExtraData))
+						w.Write(ExtraData);
+
+					break;
+				}
 
-			if (Target.SerializableType != TargetType.Invalid)
-				fields |= OrderFields.Target;
+				case OrderType.World:
```

Comment:
If we move `Subject` into an optional target order field we can merge `OrderType.Immediate` and `OrderType.World` into a single `OrderType.Standard`. We already use the order's frame in most places to distinguish whether it is immediate or not.

This then saves us from implementing a new order type.

Reply:
Proof of concept: https://github.com/pchote/OpenRA/commits/order-protocol-2

Reply:
Updated.